File: DQNAgent.js
class DQNAgent {
    discount = 0.99;
    replayMemorySize = 50000;
    minReplayMemorySize = 1000;
    minibatchSize = 64;
    updateTargetEvery = 5;

    constructor() {
        // Main model
        this.model = this.createModel();
        // Target model
        this.targetModel = this.createModel();
        this.targetModel.setWeights(this.model.getWeights());
        // Memory buffer
        this.replayMemory = [];
        // Counter for when to update target network
        this.targetUpdateCounter = 0;

    }

    createModel() {
        const model = tf.sequential();
        model.add(tf.layers.conv2d({
            inputShape: env.observationSpaceValues,
            kernelSize: 3,
            filters: 256,
            strides: 1,
            activation: 'relu'
        }));
        model.add(tf.layers.maxPooling2d({poolSize: 2}));
        model.add(tf.layers.dropout({rate: 0.2}));

        model.add(tf.layers.conv2d({
            kernelSize: 3,
            filters: 256,
            strides: 1,
            activation: 'relu'
        }));
        model.add(tf.layers.maxPooling2d({poolSize: 2}));
        model.add(tf.layers.dropout({rate: 0.2}));

        model.add(tf.layers.flatten());
        model.add(tf.layers.dense({units: 64}));
        model.add(tf.layers.dense({
            units: env.actionSpaceSize,
            activation: 'linear'
        }));

        const optimizer = tf.train.adam({learningRate: 0.001});
        model.compile({
            optimizer: optimizer,
            loss: 'meanSquaredError',
            metrics: ['accuracy']
        });
        return model;
    }

    updateReplayMemory(stepResult) {
        // stepResult = {obs, action, reward, newObs, done}
        this.replayMemory.push(stepResult);
        while (this.replayMemory.length > this.replayMemorySize) {
            this.replayMemory.shift();
        }
    }

    train(terminalState) {
        if (this.replayMemory.length < this.minReplayMemorySize) {
            return;
        }

        // Randomly sampled minibatch from 
        let minibatch = _.sample(this.replayMemory, this.minibatchSize);

        // Get current states from minibatch
        let currentStatesBatch = minibatch.map(stepResult => stepResult.obs);
        // Get model predictions from current states (predicted Q values)
        // ***** maybe just predict? does tidy work?
        // ***** check for memory leaks at end
        let currentPredQsBatch = tf.tidy(() => {
            return this.model.predictOnBatch(tf.tensor(currentStatesBatch)).arraySync();
        });
        // Get model predictions for future states
        let futureStatesBatch = minibatch.map(stepResult => stepResult.newObs);
        let futurePredQsBatch = tf.tidy(() => {
            return this.model.predictOnBatch(tf.tensor(futureStatesBatch)).arraySync();
        });
        // Features and labels
        let X = [];
        let y = [];
        // Go through each step in minibatch
        for (let i = 0; i < this.minibatchSize; i++) {
            let {obs, action, reward, newObs, done} = minibatch[i];
            // Calculate new Q value
            let newQ;
            if (done) {
                newQ = reward;
            } else {
                let maxFutureQ = Math.max(...futurePredQsBatch[i]);
                // Bellman equation
                newQ = reward + this.discount * maxFutureQ;
            }
            // Update Q value for current state
            let currentPredQs = currentPredQsBatch[i];
            currentPredQs[action] = newQ;
            // Add to training data
            X.push(obs);
            y.push(currentPredQs);
        }
        // Fit on minibatch
        tf.tidy(() => {
            this.model.fit(tf.tensor(X), tf.tensor(y), {
                batchSize: this.minibatchSize,
                epochs: 1,
                verbose: 0,
                shuffle: false
            });
        });
        
        // Update target network counter every episode
        if (terminalState) {
            this.targetUpdateCounter++;
        }

        // Update target network
        if (this.targetUpdateCounter > this.updateTargetEvery) {
            this.targetModel.setWeights(this.model.getWeights());
            this.targetUpdateCounter = 0;
        }
    }

    getQs(obs) {
        return tf.tidy(() => {
            return this.model.predict(tf.tensor(obs).reshape([-1, ...env.observationSpaceValues])).arraySync()[0];
        });
    }
}

File: DQN_agent.py
import numpy as np
from collections import deque
import random
import time

import tensorflow as tf
from tensorflow.keras.optimizers import Adam
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense, Dropout, Conv2D, MaxPooling2D, Flatten
from tensorflow.keras.callbacks import Callback

from snake_env import SnakeEnv

gpus = tf.config.experimental.list_physical_devices("GPU")
if gpus:
    try:
        for gpu in gpus:
            tf.config.experimental.set_memory_growth(gpu, True)
    except RuntimeError as e:
        print(e)

class DQNAgent:
    MODEL_NAME = "128_256v2"
    LEARNING_RATE = 1e-3
    DISCOUNT = 0.99
    REPLAY_MEMORY_SIZE = 50_000
    MIN_MEMORY_SIZE = 1_000
    MINIBATCH_SIZE = 64
    UPDATE_TARGET_EVERY = 5

    def __init__(self):
        # Main model
        self.model = self.create_model()
        # Target model
        self.target_model = self.create_model()
        self.target_model.set_weights(self.model.get_weights())
        # Memory buffer
        self.replay_memory = deque(maxlen=self.REPLAY_MEMORY_SIZE)
        # Modified TensorBoard
        self.callback = CustomTensorBoard(log_dir=f"logs/{self.MODEL_NAME}-{time.time()}")
        # Counter for when to update target network
        self.target_update_counter = 0

    def create_model(self):
        model = Sequential()

        model.add(Conv2D(128, (3, 3), activation="relu", input_shape=SnakeEnv.OBSERVATION_SPACE_VALUES))
        model.add(MaxPooling2D(pool_size=(2, 2)))
        model.add(Dropout(0.2))

        model.add(Conv2D(256, (3, 3), activation="relu"))
        model.add(MaxPooling2D(pool_size=(2, 2)))
        model.add(Dropout(0.2))

        model.add(Flatten())
        model.add(Dense(64))
        model.add(Dense(SnakeEnv.ACTION_SPACE_SIZE, activation="linear"))
        
        model.compile(loss="mse", optimizer=Adam(learning_rate=self.LEARNING_RATE), metrics=["accuracy"])
        return model

    def update_replay_memory(self, step_result):
        # stepResult = (obs, action, reward, newObs, done)
        self.replay_memory.append(step_result)

    def train(self, terminalState):
        if len(self.replay_memory) < self.MIN_MEMORY_SIZE:
            return

        # Randomly sampled minibatch from 
        minibatch = random.sample(self.replay_memory, self.MINIBATCH_SIZE)

        # Get current states from minibatch
        current_obs_mb = np.array([step_results[0] for step_results in minibatch])/255
        # Get model predictions from current states (predicted Q values)
        current_qs_mb = self.model.predict(current_obs_mb)

        # Get model predictions for future states
        future_obs_mb = np.array([step_results[3] for step_results in minibatch])/255
        future_qs_mb = self.target_model.predict(future_obs_mb)

        # Features and labels
        X = []
        y = []

        # Go through each step in minibatch
        for i, (obs, action, reward, newObs, done) in enumerate(minibatch):
            # Calculate new Q value
            if not done:
                max_future_q = np.max(future_qs_mb[i])
                new_q = reward + self.DISCOUNT * max_future_q
            else:
                new_q = reward

            # Update Q value for current state
            current_qs = current_qs_mb[i]
            current_qs[action] = new_q

            # Add to training data
            X.append(obs)
            y.append(current_qs);

        # Fit on minibatch
        self.model.fit(np.array(X)/255, np.array(y), batch_size=self.MINIBATCH_SIZE, 
                        verbose=0, shuffle=False, callbacks=[self.callback] if terminalState else None)
        
        # Update target network counter every episode
        if terminalState:
            self.target_update_counter += 1

        # Update target network
        if self.target_update_counter > self.UPDATE_TARGET_EVERY:
            self.target_model.set_weights(self.model.get_weights())
            self.target_update_counter = 0

    def get_qs(self, obs):
        return self.model.predict(obs.reshape(-1, *SnakeEnv.OBSERVATION_SPACE_VALUES)/255)[0]

class CustomTensorBoard(Callback):
    def __init__(self, log_dir):
        self.step = 1
        self.writer = tf.summary.create_file_writer(log_dir)

    def update_stats(self, **logs):
        with self.writer.as_default():
            for name, value in logs.items():
                tf.summary.scalar(name, value, self.step)
        self.writer.flush()
